src/plugins/data/public/ui/dataset_selector/dataset_selector.tsx

```diff
@@ -57,15 +58,26 @@ export const DatasetSelector = ({
 
     if (!isMounted.current) return;
 
+    const recentDatasets = datasetService.getRecentDatasets();
     const fetchedDatasets =
       fetchedIndexPatternDataStructures.children?.map((pattern) =>
         typeConfig.toDataset([pattern])
       ) ?? [];
-    setDatasets(fetchedDatasets);
+    const filteredFetchedDatasets = fetchedDatasets.filter(
+      (dataset) => !recentDatasets.some((ds) => ds.id === dataset.id)
+    );
+    const allDatasets = [
+      ...recentDatasets.map((ds) => ({ dataset: ds, recent: true })),
+      ...filteredFetchedDatasets.map((ds) => ({ dataset: ds, recent: false })),
+    ];
+    setDatasets(allDatasets);
 
     // If no dataset is selected, select the first one
     if (!selectedDataset && fetchedDatasets.length > 0) {
-      setSelectedDataset(fetchedDatasets[0]);
+      setSelectedDataset(
+        allDatasets.filter((ds) => ds.recent)[0].dataset ??
+          allDatasets.filter((ds) => !ds.recent)[0].dataset
+      );
```

Comment:
is the second filter needed? since if there's no recent datasets, then every dataset should satisfy `!ds.recent`

also recent datasets are added at first anyways, this is just `allDatasets[0].dataset`?